### PR TITLE
bpo-41324 Add a minimal decimal capsule API

### PR DESCRIPTION
**Pull request by skrah**
https://bugs.python.org/issue41324

### CODE REVIEW

**Comment by pitrou** [comment]
Also cc @xhochy just in case he's interested in this.

**Comment by pitrou** [comment]
Thank you @skrah for this. The API here would be enough for explicitly-typed conversion.

For type inference (I'm not sure that would make sense, but perhaps) there would ideally be an additional API to get the number of digits in a Decimal (that is, `len(dec.as_tuple().digits)`).

File: Include/decimal.h [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+/*                               Libmpdec API                               */
+/****************************************************************************/
+
+/*
+ * Copied from libmpdec_triple.h. This will be in mpdecimal.h once the
+ * functions are integrated into libmpdec.
+ */
+
+#ifndef LIBMPDEC_TRIPLE_H_
+struct mpd_t;
```
Comment (pitrou) [inline Include/decimal.h]: Is the `mpd_t` struct ABI-stable? It might be worth mentioning it here.

Comment (skrah) [reply Include/decimal.h]: Yes, I have no plans to change it.   If some spectacular feature emerges that would justify a change I'd put it into `libmpdec-3.0` and start a new series.  But it's very unlikely.

**Comment by skrah** [comment]
> For type inference (I'm not sure that would make sense, but perhaps) there would ideally be an additional API to get the number of digits in a Decimal (that is, `len(dec.as_tuple().digits)`).

That's a good point.  Initially I thought people would use `PyDec_Get(v)->digits`, but then you need to include
`mpdecimal.h`.  So actually we have two APIs: One for the people who use `libmpdec` directly and the easier
one.  It would make sense to add `PyDec_GetDigits(v)`.

File: Doc/c-api/decimal.rst [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+
+Initialize
+----------
+
+Typically, a C extension module that uses the decimal API will do these
+steps in its init function:
+
+.. code-block::
+
+    #include <decimal.h>
```
Comment (pitrou) [inline Doc/c-api/decimal.rst]: Is "decimal.h" really reachable at the toplevel of the include path? If so, isn't there a risk of name clash?

Comment (skrah) [reply Doc/c-api/decimal.rst]: Hmm, I copied what `datetime.h` is doing: It is in `Include/datetime.h` but not included in `Python.h`.  It is copied to e.g. `$prefix/include/python3.8m/datetime.h`, so I suppose distutils finds that.

It should probably be `#include "decimal.h"`, I'll check for a real  external extension module.

Comment (skrah) [reply Doc/c-api/decimal.rst]: The name `decimal.h` can certainly clash, but so can `datetime.h`.

Comment (pitrou) [reply Doc/c-api/decimal.rst]: Indeed, I have no strong feelings about it. That said, it looks like there are [a number of projects](https://github.com/search?l=&q=filename%3Adecimal.h&type=Code) with include files named `decimal.h`, and there are already [complaints](https://bugreports.qt.io/browse/QTBUG-29985) about existing conflicts.

Perhaps naming it e.g. `pydecimal.h` would help avoid any ambiguities.

Comment (skrah) [reply Doc/c-api/decimal.rst]: Yes, that's probably safer.  I've renamed it to `pydecimal.h`, the instructions work with an installed `Python`.

**Pull request merged**